src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.Razor;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    /// <summary>
+    /// Executing a Razor Page requires ensuring it's compiled which is an asynchronous operation. 
+    /// The <see cref="IActionInvokerProvider"/> pipeline is synchronous and consequently we cannot compile a Razor Page
+    /// as part of executing it without blocking. The <see cref="PageActionInvokerProvider"/> instead delegates to this type
+    /// to compile the Razor Page and cache the information required for page invocation.
+    /// </summary>
+    internal class CreatePageCacheEntryActionInvoker : IActionInvoker
```

Comment:
What's the benefit of having this as a separate class? What if the default invoker for pages just did this.